sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp

```diff
@@ -593,367 +586,529 @@ pi_result piContextCreate(const pi_context_properties *Properties,
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
 pi_result piKernelGetInfo(pi_kernel, pi_kernel_info, size_t, void *, size_t *) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
 pi_result piKernelGetGroupInfo(pi_kernel, pi_device, pi_kernel_group_info,
                                size_t, void *, size_t *) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
-pi_result
-piKernelGetSubGroupInfo(pi_kernel, pi_device,
-                        pi_kernel_sub_group_info, // TODO: untie from OpenCL
-                        size_t, const void *, size_t, void *, size_t *) {
+pi_result piKernelGetSubGroupInfo(pi_kernel, pi_device,
+                                  pi_kernel_sub_group_info, size_t,
+                                  const void *, size_t, void *, size_t *) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
 pi_result piKernelRetain(pi_kernel) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
 pi_result piKernelRelease(pi_kernel) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
 pi_result piEventCreate(pi_context, pi_event *) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
 pi_result piEventGetInfo(pi_event, pi_event_info, size_t, void *, size_t *) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
 pi_result piEventGetProfilingInfo(pi_event Event, pi_profiling_info ParamName,
                                   size_t ParamValueSize, void *ParamValue,
                                   size_t *ParamValueSizeRet) {
-  (void)Event;
-  (void)ParamName;
-  (void)ParamValueSize;
-  (void)ParamValue;
-  (void)ParamValueSizeRet;
-  DIE_NO_IMPLEMENTATION;
+  std::cerr << "Warning : Profiling Not supported under PI_ESIMD_CPU"
+            << std::endl;
   return PI_SUCCESS;
 }
 
 pi_result piEventsWait(pi_uint32 NumEvents, const pi_event *EventList) {
-  (void)NumEvents;
-  (void)EventList;
-  DIE_NO_IMPLEMENTATION;
+  for (int i = 0; i < (int)NumEvents; i++) {
+    if (EventList[i]->IsDummyEvent) {
+      continue;
+    }
+    int Result = EventList[i]->CmEventPtr->WaitForTaskFinished();
+    if (Result != cm_support::CM_SUCCESS) {
+      return PI_OUT_OF_RESOURCES;
+    }
+  }
   return PI_SUCCESS;
 }
 
 pi_result piEventSetCallback(pi_event, pi_int32,
                              void (*)(pi_event, pi_int32, void *), void *) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
 pi_result piEventSetStatus(pi_event, pi_int32) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
 pi_result piEventRetain(pi_event) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
 pi_result piEventRelease(pi_event Event) {
-  (void)Event;
-  DIE_NO_IMPLEMENTATION;
+  if (!Event->IsDummyEvent) {
+    if ((Event->CmEventPtr == nullptr) || (Event->OwnerQueue == nullptr)) {
+      return PI_INVALID_EVENT;
+    }
+    int Result = Event->OwnerQueue->DestroyEvent(Event->CmEventPtr);
+    if (Result != cm_support::CM_SUCCESS) {
+      return PI_INVALID_EVENT;
+    }
+  }
+  delete Event;
```

Comment:
Add TODO the this is only safe when piEventRetain is `DIE_NO_IMPLEMENTATION`. I wonder why not implement ref counting right away to avoid risks in future.